### PR TITLE
Display additional messaging around the impact of running Firebase extensions with nodejs10 runtime.

### PR DESCRIPTION
**Pull request by taeold**
Deploying Cloud Functions with nodejs10 runtime has new billing behaviors that we want Firebase Extensions users to be aware of.

This change adds additional messaging for the users in following 4 scenarios:

| | No Billing | Billing (Updated) |
| - | - | - |
| Install | install-nobilling | Screen Shot 2020-08-20 at 3 58 55 PM |
| Update | X |  Screen Shot 2020-08-20 at 3 59 18 PM |

### CODE REVIEW

**Comment by taeold** [comment]
@joehan @laurenzlong Could you guys take a look at the code structure? The exact messages are still being finalized.

File: src/extensions/nodejsMigrationHelper.ts [hunk 1]
```diff
@@ -0,0 +1,75 @@
...
+const changeMsg =
+  "Node.js 8 has been deprecated. It’s recommended to update this extension to ensure it is running on Node.js 10.";
+const billingMsgUpdate =
+  "This update includes an upgrade to Node.js 10 from Node.js 8, which is no longer maintained. Your project must be on the Blaze (pay as you go) plan to deploy Node.js 10 functions. Starting with this update, you will be charged a small amount when you deploy this extension, including when you make configuration changes and apply future updates.";
+const billingMsgCreate =
+  "This extension includes Node.js 10 functions. Your project must be on the Blaze (pay as you go) plan to deploy Node.js 10 functions. You will be charged a small amount when you deploy this extension, including when you make configuration changes and apply future updates.";
+
+function hasRuntime(resource: extensionsApi.Resource, runtime: string): boolean {
+  let resourceRuntime = "nodejs8"; // default to nodejs8 runtime
+  if (resource.properties && resource.properties.runtime) {
```
Comment (joehan) [inline src/extensions/nodejsMigrationHelper.ts:16]: Prefer using optional chaining here (https://www.typescriptlang.org/docs/handbook/release-notes/typescript-3-7.html): 
`if (resource?.properties?.runtime){`

File: src/extensions/nodejsMigrationHelper.ts [hunk 2]
```diff
@@ -0,0 +1,75 @@
...
+}
+
+function shouldDisplayMsg(
+  newSpec: extensionsApi.ExtensionSpec,
+  curSpec?: extensionsApi.ExtensionSpec
+): boolean {
+  const newResources = newSpec.resources;
+  const curResources = curSpec?.resources || [];
+  return (
+    newResources.some((r) => hasRuntime(r, "nodejs10")) &&
```
Comment (joehan) [inline src/extensions/nodejsMigrationHelper.ts:29]: Optional: There is an edge case here, where if someone made an extension with a mix of node8 and node 10 functions, it would always show this message on updates, even if no runtimes changed. I don't think this is a meaningful edge case or requires change, but if you can think of a nice way to handle it, we may as well.

Comment (joehan) [inline src/extensions/nodejsMigrationHelper.ts:29]: This statement is a bit dense, and I think this would be a bit more readable if we split it into a couple variables:
```
const newResourcesUseNode10 = newResources.some((r) => hasRuntime(r, "nodejs10"));
const curResourcesUseNode8 = curResources.some((r) => hasRuntime(r, "nodejs8"));
return (newResourcesUseNode10 && (!curSpec || curResourcesUseNode8))
```

File: src/test/extensions/nodejsMigrationHelper.spec.ts [hunk 3]
```diff
@@ -0,0 +1,122 @@
...
+    },
+  ],
+  author: { authorName: "Tester" },
+  contributors: [{ authorName: "Tester 2" }],
+  billingRequired: true,
+  sourceUrl: "test.com",
+  params: [],
+};
+
+describe("nodejsMigrationHelper", () => {
```
Comment (joehan) [inline src/test/extensions/nodejsMigrationHelper.spec.ts:71]: Optional: Add a test case for if the user says no to the prompt

File: src/extensions/nodejsMigrationHelper.ts [hunk 4]
```diff
@@ -0,0 +1,75 @@
+import { FirebaseError } from "../error";
+import * as extensionsApi from "./extensionsApi";
+import { logPrefix } from "./extensionsHelper";
+import { promptOnce } from "../prompt";
+import * as utils from "../utils";
+
+const changeMsg =
+  "Node.js 8 has been deprecated. It’s recommended to update this extension to ensure it is running on Node.js 10.";
```
Comment (karayu) [inline src/extensions/nodejsMigrationHelper.ts:8]: For updates: This feels weird to me as a warning in the beginning of the update process because it's telling you to go do something but you're already doing it. @kevinthecheung wondering if it makes sense to combine with the billingMsgUpdate?

File: src/extensions/nodejsMigrationHelper.ts (hunk 1 above)
Comment (taeold) [reply src/extensions/nodejsMigrationHelper.ts:16]: Great call.

File: src/extensions/nodejsMigrationHelper.ts (hunk 2 above)
Comment (taeold) [reply src/extensions/nodejsMigrationHelper.ts:29]: Ack. I don't think we'll ever see the edge case you described while we have control over all published extensions.

Comment (taeold) [reply src/extensions/nodejsMigrationHelper.ts:29]: Agreed!

File: src/test/extensions/nodejsMigrationHelper.spec.ts (hunk 3 above)
Comment (taeold) [reply src/test/extensions/nodejsMigrationHelper.spec.ts:72]: Done

File: src/extensions/nodejsMigrationHelper.ts [hunk 5]
```diff
@@ -0,0 +1,74 @@
...
+const billingMsgUpdate =
+  "This update includes an upgrade to Node.js 10 from Node.js 8, which is no longer maintained. Your project must be on the Blaze (pay as you go) plan to deploy Node.js 10 functions. Starting with this update, you will be charged a small amount when you deploy this extension, including when you make configuration changes and apply future updates.";
```
Comment (kevinthecheung) [inline src/extensions/nodejsMigrationHelper.ts:10]:
```suggestion
  "This update includes an upgrade to Node.js 10 from Node.js 8, which is no longer maintained. Node.js 10 requires your project to be on the Blaze (pay as you go) plan, and starting with this update, you will be charged a small amount (typically around $0.01/month) for each instance of this extension you've installed, in addition to any charges associated with your use of Firebase services.";
```

File: src/extensions/nodejsMigrationHelper.ts [hunk 6]
```diff
@@ -0,0 +1,74 @@
...
+const billingMsgCreate =
+  "This extension includes Node.js 10 functions. Your project must be on the Blaze (pay as you go) plan to deploy Node.js 10 functions. You will be charged a small amount when you deploy this extension, including when you make configuration changes and apply future updates.";
```
Comment (kevinthecheung) [inline src/extensions/nodejsMigrationHelper.ts:12]:
```suggestion
  "This extension includes Node.js 10 functions. Node.js 10 requires your project to be on the Blaze (pay as you go) plan, and you will be charged a small amount (typically around $0.01/month) for each instance of this extension you've installed, in addition to any charges associated with your use of Firebase services.";
```

File: src/extensions/nodejsMigrationHelper.ts (hunk 4 above)
Comment (kevinthecheung) [reply src/extensions/nodejsMigrationHelper.ts:8]: Maybe we don't need `displayNodejsChangeNotice()`?

Comment (taeold) [reply src/extensions/nodejsMigrationHelper.ts:8]: Removed!

Comment (karayu) [reply src/extensions/nodejsMigrationHelper.ts]: LGTM! I think we're all set as long as Annie is good.

File: src/extensions/nodejsMigrationHelper.ts (hunk 6 above)
Comment (karayu) [reply src/extensions/nodejsMigrationHelper.ts]: LGTM! I think we're all set as long as Annie is good.

**Comment by annie-ryan** [comment]
@karayu @kevinthecheung for an FYI on this comment

> 1. When _installing_ extensions w/ nodejs10 runtime
> 
> ```
> ⚠  extensions: This extension includes Node.js 10 functions. Your project must be on the Blaze (pay as you go) plan to deploy Node.js 10 functions. You will be charged a small amount when you deploy this extension, including when you make configuration changes and apply future updates.
> ```

**CONSOLE/CLI COPY CHECK. For this journey in the console, LGTM'd copy reads:**
"You will be charged around $0.01/month for each instance of this extension you install. Additionally, using this extension will contribute to your project's overall usage level of Firebase services. However, you'll only be charged for usage that exceeds Firebase's free tier for those services. Learn more details in the Firebase pricing page"

We decided to not be explicit about Node 10 for _installing_ journeys, so the copy won't grow stale once extensions are on Node 10+

"$0.01/month" is an inline link to the (in progress) pricing sample
"Learn more details in the Firebase pricing page" is an inline link to firebase.google.com/pricing

Could we add these links to the end of the callout paragraph in the CLI?

> 1. When _updating_ extensions from nodejs8 to nodejs10 runtime
> 
> ```
> ⚠  extensions: Node.js 8 has been deprecated. It’s recommended to update this extension to ensure it is running on Node.js 10.
> ⚠  extensions: This extension includes Node.js 10 functions. Your project must be on the Blaze (pay as you go) plan to deploy Node.js 10 functions. You will be charged a small amount when you deploy this extension, including when you make configuration changes and apply future updates.
>```

**CONSOLE/CLI COPY CHECK. For the FIRST callout, LGTM'd copy reads:**
"This update upgrades the extension's runtime to Node.js 10 from Node.js 8, which is no longer maintained. Install this update to ensure it's using a supported runtime.
- Starting February 15, 2021 if you haven't updated this extension, you will no longer be able to deploy configuration changes.
- Starting March 15, 2021 if you haven't updated this extension, it will no longer work.

Note that upgrading to Node.js 10 has some billing implications. Review the billing and usage information below."

**CONSOLE/CLI COPY CHECK. For the SECOND callout, LGTM'd copy reads:**
"This update includes an upgrade to Node.js 10 from Node.js 8, which is no longer maintained. Starting with this update, you will be charged a small amount (typically around $0.01/month) for each instance of this extension you've installed, in addition to any charges associated with your use of Firebase services."

In the console, we have 2 buttons after this text:
"See pricing examples" - links to the (in progress) pricing sample
"Read FAQ" - links to firebase.google.com/support/faq/#functions-pricing

Could we add these links to the end of the callout paragraph in the CLI?

**Comment by annie-ryan** [comment]
@taeold also, I haven't seen in this PR what the CLI is printing out when the user has to **upgrade** their billing plan for _installing_ and _updating_ ?

@karayu @kevinthecheung

File: src/extensions/nodejsMigrationHelper.ts [hunk 7]
```diff
@@ -0,0 +1,57 @@
...
+    resourceRuntime = resource.properties.runtime;
+  }
+  return resourceRuntime == runtime;
+}
+
+function shouldDisplayMsg(
+  newSpec: extensionsApi.ExtensionSpec,
+  curSpec?: extensionsApi.ExtensionSpec
+): boolean {
+  const newResources = newSpec.resources;
```
Comment (laurenzlong) [inline src/extensions/nodejsMigrationHelper.ts:24]: is there a reason why you didn't write this as:
`const newResources = newSpec.resources || [];`

File: src/extensions/nodejsMigrationHelper.ts [hunk 8]
```diff
@@ -0,0 +1,57 @@
...
+ */
+export async function displayNodejsBillingNotice(
+  newSpec: extensionsApi.ExtensionSpec,
+  curSpec?: extensionsApi.ExtensionSpec,
+  update?: boolean
+): Promise<void> {
+  if (shouldDisplayMsg(newSpec, curSpec)) {
+    const msg = update ? billingMsgUpdate : billingMsgCreate;
+
+    utils.logLabeledWarning(logPrefix, msg);
```
Comment (laurenzlong) [inline src/extensions/nodejsMigrationHelper.ts:47]: @annie-ryan @karayu do we actually want this as a warning? (Will display with orange text). We usually reserve that for when the user has done something wrong, or if it's really important that they pay attention. I think info level would be better (blue text)

Comment (laurenzlong) [reply src/extensions/nodejsMigrationHelper.ts:47]: @taeold It would be helpful if you can screenshot instead of copy/paste in the future (since the reviewers can then see colors, bolding, etc.)

File: src/extensions/nodejsMigrationHelper.ts [hunk 9]
```diff
@@ -0,0 +1,57 @@
...
+    const continueUpdate = await promptOnce({
+      type: "confirm",
+      message: "Do you wish to continue?",
+      default: false,
```
Comment (laurenzlong) [inline src/extensions/nodejsMigrationHelper.ts:51]: Default to true, otherwise it gets annoying since people will often just click "enter" and then accidentally exit the flow.

File: src/extensions/nodejsMigrationHelper.ts [hunk 10]
```diff
@@ -0,0 +1,57 @@
...
+ *
+ * @param newSpec A extensionSpec to compare to
+ * @param curSpec A current extensionSpec
+ * @param update Display update message if true
+ * @return Displayed message
+ */
+export async function displayNodejsBillingNotice(
+  newSpec: extensionsApi.ExtensionSpec,
+  curSpec?: extensionsApi.ExtensionSpec,
+  update?: boolean
```
Comment (laurenzlong) [inline src/extensions/nodejsMigrationHelper.ts:42]: I don't see this "update" parameter used when ext-update.ts is calling this method. Also I understand that you're trying to prevent repeated code so you have a general method for both update and create. I think this actually makes the code harder to reason about, and the shouldDisplayMsg method is rather convoluted. What do you think about having 2 separate methods: `displayNodejsUpdateBillingNotice` and `displayNodejsCreateBillingNotice`, and have these 2 lines as global variables:
- const newResourcesUseNode10 = newResources.some((r) => hasRuntime(r, "nodejs10"));
- const curResourcesUseNode8 = curResources.some((r) => hasRuntime(r, "nodejs8"));

Then the logic for figuring out whether or not to display the message can be directly inside `displayNodejsUpdateBillingNotice` and `displayNodejsCreateBillingNotice` methods. That will make this file easier to read and less error prone.

**Comment by karayu** [comment]
Annie has great points and matching console sounds good in terms of not calling out node 10. @kevinthecheung do you mind taking a look and adapting for cli? I wonder if it needs to be a shortened version?

File: src/extensions/nodejsMigrationHelper.ts [hunk 11]
```diff
@@ -0,0 +1,57 @@
...
+import { promptOnce } from "../prompt";
+import * as utils from "../utils";
+
+const billingMsgUpdate =
+  "This update includes an upgrade to Node.js 10 from Node.js 8, which is no longer maintained. Node.js 10 requires your project to be on the Blaze (pay as you go) plan, and starting with this update, you will be charged a small amount (typically around $0.01/month) for each instance of this extension you've installed, in addition to any charges associated with your use of Firebase services.";
+const billingMsgCreate =
+  "This extension includes Node.js 10 functions. Node.js 10 requires your project to be on the Blaze (pay as you go) plan, and you will be charged a small amount (typically around $0.01/month) for each instance of this extension you've installed, in addition to any charges associated with your use of Firebase services.";
+
+function hasRuntime(resource: extensionsApi.Resource, runtime: string): boolean {
+  let resourceRuntime = "nodejs8"; // default to nodejs8 runtime
```
[the remark this answers is not in the document]
Comment (laurenzlong) [reply src/extensions/nodejsMigrationHelper.ts:13]: We may change the default runtime in future spec versions.

File: src/extensions/nodejsMigrationHelper.ts [hunk 12]
```diff
@@ -0,0 +1,57 @@
+import { FirebaseError } from "../error";
+import * as extensionsApi from "./extensionsApi";
+import { logPrefix } from "./extensionsHelper";
+import { promptOnce } from "../prompt";
+import * as utils from "../utils";
+
+const billingMsgUpdate =
+  "This update includes an upgrade to Node.js 10 from Node.js 8, which is no longer maintained. Node.js 10 requires your project to be on the Blaze (pay as you go) plan, and starting with this update, you will be charged a small amount (typically around $0.01/month) for each instance of this extension you've installed, in addition to any charges associated with your use of Firebase services.";
+const billingMsgCreate =
+  "This extension includes Node.js 10 functions. Node.js 10 requires your project to be on the Blaze (pay as you go) plan, and you will be charged a small amount (typically around $0.01/month) for each instance of this extension you've installed, in addition to any charges associated with your use of Firebase services.";
```
Comment (kevinthecheung) [inline src/extensions/nodejsMigrationHelper.ts:10]:
```suggestion
  "You will be charged around $0.01/month for each instance of this extension"
  + " you install. Additionally, using this extension will contribute to your"
  + " project's overall usage level of Firebase services. However, you'll only"
  + " be charged for usage that exceeds Firebase's free tier for those"
  + " services.\n\n"
  + "See pricing examples: https://TBD\n"
  + "Learn about Firebase pricing: https://firebase.google.com/pricing";
```

File: src/extensions/nodejsMigrationHelper.ts [hunk 13]
```diff
@@ -0,0 +1,57 @@
+import { FirebaseError } from "../error";
+import * as extensionsApi from "./extensionsApi";
+import { logPrefix } from "./extensionsHelper";
+import { promptOnce } from "../prompt";
+import * as utils from "../utils";
+
+const billingMsgUpdate =
+  "This update includes an upgrade to Node.js 10 from Node.js 8, which is no longer maintained. Node.js 10 requires your project to be on the Blaze (pay as you go) plan, and starting with this update, you will be charged a small amount (typically around $0.01/month) for each instance of this extension you've installed, in addition to any charges associated with your use of Firebase services.";
```
Comment (kevinthecheung) [inline src/extensions/nodejsMigrationHelper.ts:8]:
```suggestion
  "This update includes an upgrade to Node.js 10 from Node.js 8, which is no"
  + " longer maintained. Starting with this update, you will be charged a"
  + " small amount (typically around $0.01/month) for each instance of this"
  + " extension you've installed, in addition to any charges associated with"
  + " your use of Firebase services.\n\n"
  + "See pricing examples: https://TBD\n"
  + "See the FAQ: https://firebase.google.com/support/faq/#extensions-pricing";
```

**Comment by kevinthecheung** [comment]
> **CONSOLE/CLI COPY CHECK. For the FIRST callout, LGTM'd copy reads:**
> "This update upgrades the extension's runtime to Node.js 10 from Node.js 8, which is no longer maintained. Install this update to ensure it's using a supported runtime.
> 
> * Starting February 15, 2021 if you haven't updated this extension, you will no longer be able to deploy configuration changes.
> * Starting March 15, 2021 if you haven't updated this extension, it will no longer work.
> 
> Note that upgrading to Node.js 10 has some billing implications. Review the billing and usage information below."

I think we decided this callout isn't necessary in the CLI since the developer is already doing the upgrade.

**Comment by annie-ryan** [comment]
> > **CONSOLE/CLI COPY CHECK. For the FIRST callout, LGTM'd copy reads:**
> > "This update upgrades the extension's runtime to Node.js 10 from Node.js 8, which is no longer maintained. Install this update to ensure it's using a supported runtime.
> > 
> > * Starting February 15, 2021 if you haven't updated this extension, you will no longer be able to deploy configuration changes.
> > * Starting March 15, 2021 if you haven't updated this extension, it will no longer work.
> > 
> > Note that upgrading to Node.js 10 has some billing implications. Review the billing and usage information below."
> 
> I think we decided this callout isn't necessary in the CLI since the developer is already doing the upgrade.

Is there any place before this where they get notice of these dates? Or are we assuming they will remember them from the MSA?

File: src/extensions/nodejsMigrationHelper.ts (hunk 13 above)
[the remark this answers is not in the document]
Comment (taeold) [reply src/extensions/nodejsMigrationHelper.ts]: Done (by Kevin)!

File: src/extensions/nodejsMigrationHelper.ts (hunk 12 above)
[the remark this answers is not in the document]
Comment (taeold) [reply src/extensions/nodejsMigrationHelper.ts]: Done (by Kevin)!

File: src/extensions/nodejsMigrationHelper.ts [hunk 14]
```diff
@@ -0,0 +1,57 @@
...
+const billingMsgCreate =
+  "This extension includes Node.js 10 functions. Node.js 10 requires your project to be on the Blaze (pay as you go) plan, and you will be charged a small amount (typically around $0.01/month) for each instance of this extension you've installed, in addition to any charges associated with your use of Firebase services.";
+
+function hasRuntime(resource: extensionsApi.Resource, runtime: string): boolean {
+  let resourceRuntime = "nodejs8"; // default to nodejs8 runtime
+  if (resource.properties?.runtime) {
+    resourceRuntime = resource.properties.runtime;
+  }
+  return resourceRuntime == runtime;
```
[the remark this answers is not in the document]
Comment (taeold) [reply src/extensions/nodejsMigrationHelper.ts:29]: Dine.

File: src/extensions/nodejsMigrationHelper.ts (hunk 7 above)
Comment (taeold) [reply src/extensions/nodejsMigrationHelper.ts:36]: I assumed that resources would never be empty, but that now realized that isn't a good assumption. Fixed.

File: src/extensions/nodejsMigrationHelper.ts (hunk 9 above)
Comment (taeold) [reply src/extensions/nodejsMigrationHelper.ts:63]: Great point. Done.

File: src/extensions/nodejsMigrationHelper.ts (hunk 10 above)
Comment (taeold) [reply src/extensions/nodejsMigrationHelper.ts]: Great advice. I refactored the code based on changes you recommended here.

File: src/extensions/nodejsMigrationHelper.ts (hunk 11 above)
[the remark this answers is not in the document]
Comment (taeold) [reply src/extensions/nodejsMigrationHelper.ts]: Good to know! Done.

File: src/extensions/billingMigrationHelper.ts [hunk 15]
```diff
@@ -0,0 +1,78 @@
...
+import * as utils from "../utils";
+
+const billingMsgUpdate =
+  "This update includes an upgrade to Node.js 10 from Node.js 8, which is no" +
+  " longer maintained. Starting with this update, you will be charged a" +
+  " small amount (typically around $0.01/month) for each instance of this" +
+  " extension you've installed, in addition to any charges associated with" +
+  " your use of Firebase services.\n\n" +
+  "See pricing examples: https://TBD\n" +
+  "See the FAQ: https://firebase.google.com/support/faq/#extensions-pricing";
```
Comment (kevinthecheung) [inline src/extensions/billingMigrationHelper.ts:14]:
```suggestion
  "See the FAQ: https://firebase.google.com/support/faq/#extensions-runtime";
```

File: src/extensions/billingMigrationHelper.ts [hunk 16]
```diff
@@ -0,0 +1,78 @@
...
+import { promptOnce } from "../prompt";
+import * as utils from "../utils";
+
+const billingMsgUpdate =
+  "This update includes an upgrade to Node.js 10 from Node.js 8, which is no" +
+  " longer maintained. Starting with this update, you will be charged a" +
+  " small amount (typically around $0.01/month) for each instance of this" +
+  " extension you've installed, in addition to any charges associated with" +
+  " your use of Firebase services.\n\n" +
+  "See pricing examples: https://TBD\n" +
```
Comment (kevinthecheung) [inline src/extensions/billingMigrationHelper.ts:13]:
```suggestion
  "See pricing examples: https://cloud.google.com/functions/pricing#pricing_examples\n" +
```

File: src/extensions/billingMigrationHelper.ts [hunk 17]
```diff
@@ -0,0 +1,78 @@
...
+  "See the FAQ: https://firebase.google.com/support/faq/#extensions-pricing";
+const billingMsgCreate =
+  "You will be charged around $0.01/month for each instance of this extension" +
+  " you install. Additionally, using this extension will contribute to your" +
+  " project's overall usage level of Firebase services. However, you'll only" +
+  " be charged for usage that exceeds Firebase's free tier for those" +
+  " services.\n\n" +
+  "See pricing examples: https://TBD\n" +
```
Comment (kevinthecheung) [inline src/extensions/billingMigrationHelper.ts:21]:
```suggestion
  "See pricing examples: https://cloud.google.com/functions/pricing#pricing_examples\n" +
```

File: src/extensions/nodejsMigrationHelper.ts (hunk 8 above)
Comment (taeold) [reply src/extensions/nodejsMigrationHelper.ts]: Updated the PR description with screenshot.

Maybe it's because I'm using a basic terminal scheme, but I don't see warning texts as orange colored.

Comment (laurenzlong) [reply src/extensions/nodejsMigrationHelper.ts]: Discussed during UX weekly, this is fine to leave as-is. (Warning level is fine)

**Comment by laurenzlong** [comment]
> > > **CONSOLE/CLI COPY CHECK. For the FIRST callout, LGTM'd copy reads:**
> > > "This update upgrades the extension's runtime to Node.js 10 from Node.js 8, which is no longer maintained. Install this update to ensure it's using a supported runtime.
> > > 
> > > * Starting February 15, 2021 if you haven't updated this extension, you will no longer be able to deploy configuration changes.
> > > * Starting March 15, 2021 if you haven't updated this extension, it will no longer work.
> > > 
> > > Note that upgrading to Node.js 10 has some billing implications. Review the billing and usage information below."
> > 
> > 
> > I think we decided this callout isn't necessary in the CLI since the developer is already doing the upgrade.
> 
> Is there any place before this where they get notice of these dates? Or are we assuming they will remember them from the MSA?

As discussed during UX weekly. We are fine with not having the dates in the CLI text, since the linked FAQ has dates.

**Comment by laurenzlong** [comment]
Adding screenshots from my computer, since my terminal has the default Mac color schemes, which most people will have. 

Installing without a Blaze Plan:
Screen Shot 2020-08-19 at 12 42 30 PM

Updating without a Blaze Plan: (Note I had to update from local source, but when this is actually launched, it won't have the local source question)
Screen Shot 2020-08-19 at 12 27 15 PM

**Comment by laurenzlong** [comment]
@taeold Feedback from UX weekly, can you please modify these 2 links to be formatted the same as the Firebase (https://firebase.google.com) link? (This will probably require using the markdown rendering on those lines)
```
See pricing examples: https://cloud.google.com/functions/pricing#pricing_examples
See the FAQ: https://firebase.google.com/support/faq/#extensions-runtime
```

File: src/extensions/billingMigrationHelper.ts [hunk 18]
```diff
@@ -0,0 +1,87 @@
...
+}
+
+/**
+ * Displays billing changes if the update contains new billing requirements.
+ *
+ * @param curSpec A current extensionSpec
+ * @param newSpec A extensionSpec to compare to
+ * @param prompt If true, prompts user for confirmation
+ */
+export async function displayUpdateBillingNotice(
```
Comment (laurenzlong) [inline src/extensions/billingMigrationHelper.ts:43]: Sorry to comment on the method name again, but can you rename to `displayNode10UpdateBillingNotice` and `displayNode10CreateBillingNotice`. I know they're a bit verbose, but as currently named, it's not obvious that they're just for displaying Node 10 related billing notice. It sounds like they display general billing notices.

Comment (taeold) [reply src/extensions/billingMigrationHelper.ts:43]: No worries. Done.

**Comment by joehan** [comment]
Code changes look mostly good to me - it looks like you just need to run `npm run format` to fix the linting issues.

**Comment by taeold** [comment]
Updated the screenshots for "Billing" column with the added hyperlinks!

File: src/extensions/billingMigrationHelper.ts [hunk 19]
```diff
@@ -0,0 +1,97 @@
...
+});
+
+const urlPricingExamples = "https://cloud.google.com/functions/pricing#pricing_examples";
+const urlFAQ = "https://firebase.google.com/support/faq/#extensions-runtime";
+
+const billingMsgUpdate =
+  "This update includes an upgrade to Node.js 10 from Node.js 8, which is no" +
+  " longer maintained. Starting with this update, you will be charged a" +
+  " small amount (typically around $0.01/month) for the Firebase resources" +
+  " required by this extension even if it is not used in addition to any" +
```
Comment (kevinthecheung) [inline src/extensions/billingMigrationHelper.ts:21]:
```suggestion
  " required by this extension (even if it is not used), in addition to any" +
```

File: CHANGELOG.md [hunk 20]
```diff
@@ -0,0 +1 @@
+- Display additional info on the impact of running Firebase extensions with nodejs10 runtime. (#2549)
```
Comment (laurenzlong) [inline CHANGELOG.md:1]: Can you write "Node.js 10 runtime" instead?

Comment (taeold) [reply CHANGELOG.md]: Yep!
